velox/functions/prestosql/Arithmetic.cpp

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/prestosql/Arithmetic.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/prestosql/ArithmeticImpl.h"
+#include "velox/functions/prestosql/CheckedArithmeticImpl.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <typename T, typename Operation>
+class BaseFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    if (args[0]->isConstantEncoding() &&
+        args[1]->encoding() == VectorEncoding::Simple::FLAT) {
+      // Fast path for (flat, const).
+      auto constant = args[0]->asUnchecked<SimpleVector<T>>()->valueAt(0);
+      auto flatValues = args[1]->asUnchecked<FlatVector<T>>();
+      auto rawValues = flatValues->mutableRawValues();
+
+      auto rawResults =
+          getRawResults(rows, args[1], rawValues, resultType, context, result);
+
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        rawResults[row] = Operation::apply(constant, rawValues[row]);
+      });
+    } else if (
+        args[0]->encoding() == VectorEncoding::Simple::FLAT &&
```

Comment:
This code seems like a duplication more or less of the code in the above if. It is possible to refactor a common function ?